### PR TITLE
Change ExtendedRequest message

### PR DESCRIPTION
**Pull request by dododedodonl**
In order to implement RFC3062 from a server perspective, one needs to access the raw buffer of `ExtendedRequest.value` instead of the string representation.

This PR changes:
- `ExtendedRequest.value` getter is changed to still always return a string
- `ExtendedRequest.valueBuffer` getter is added to return the original raw buffer (or a buffer of the set string, but this is only for client side perspective)
- `ExtendedRequest.requestValue` now is a buffer when retrieved, instead of a string

### CODE REVIEW

File: lib/messages/ext_request.js [hunk 1]
```diff
@@ -44,14 +44,28 @@ Object.defineProperties(ExtendedRequest.prototype, {
...
     set: function setValue(val) {
       if (!(Buffer.isBuffer(val) || typeof (val) === 'string'))
         throw new TypeError('value must be a buffer or a string');
 
       this.requestValue = val;
     },
     configurable: false
+  },
+  valueBuffer: {
+    get: function getValue() {
```
Comment (UziTech) [inline lib/messages/ext_request.js]:
```suggestion
    get: function getValueBuffer() {
```

File: lib/messages/ext_request.js [hunk 2]
```diff
@@ -44,14 +44,28 @@ Object.defineProperties(ExtendedRequest.prototype, {
...
+      if(! Buffer.isBuffer(this.requestValue))
+        Buffer.fromString(this.requestValue, 'utf8');
```
Comment (UziTech) [inline lib/messages/ext_request.js]:
```suggestion
        return Buffer.fromString(this.requestValue, 'utf8');
```

**Comment by UziTech** [comment]
Why does `ExtendedRequest.requestValue` need to be a Buffer?

**Comment by dododedodonl** [comment]
In rfc2251, `ExtendedRequest.requestValue` is specified as follow:
```
     ExtendedRequest ::= [APPLICATION 23] SEQUENCE {
                requestName      [0] LDAPOID,
                requestValue     [1] OCTET STRING OPTIONAL }
```

But in rfc3062, the request is specified as follow:
```
passwdModifyOID OBJECT IDENTIFIER ::= 1.3.6.1.4.1.4203.1.11.1

PasswdModifyRequestValue ::= SEQUENCE {
     userIdentity    [0]  OCTET STRING OPTIONAL
     oldPasswd       [1]  OCTET STRING OPTIONAL
     newPasswd       [2]  OCTET STRING OPTIONAL }
```

> 2.1 A Password Modify request is an `ExtendedRequest` with the `requestName`
   field containing `passwdModifyOID` OID and optionally provides a
   requestValue field.  If the `requestValue` field is provided, it SHALL
   contain a `PasswdModifyRequestValue` with one or more fields present.

The `requestValue` is changed from `OCTET STRING` to a `SEQUENCE`. In order to parse this correctly, one needs access to the buffer. To be flexible for other extensions, the functionality is not implemented in the extended request file. An example implementation could be:

```js
var asn1 = require('asn1');
var BerReader = asn1.BerReader;

// buffer is an instance of Buffer
function passwdModifyRequestValue(buffer) {
  var requestValue = {
    userIdentity: null,
    oldPasswd: null,
    newPasswd: null,
  };

  if(! Buffer.isBuffer(buffer))
    return requestValue;

  var ber = new BerReader(buffer);

  if(ber.peek() == 0x30) {
    // Sequence
    ber.readSequence(0x30);

    if(ber.peek() == 0x80) {
      requestValue.userIdentity = ber.readString(0x80);
    }

    if(ber.peek() == 0x81) {
      requestValue.oldPasswd = ber.readString(0x81);
    }

    if(ber.peek() == 0x82) {
      requestValue.newPasswd = ber.readString(0x82);
    }
  }

  return requestValue;
}


var server = ldap.createServer();

server.exop('1.3.6.1.4.1.4203.1.11.1', function(req, res, next) {
  var parsedValue = passwordModifyRequestValue(req.valueBuffer);

  // change the password parsedValue.newPasswd in the backend

  res.end();
  return next();
});
```

**Comment by UziTech** [comment]
So for a `PasswdModifyRequestValue` would there be an issue with converting it to a string for `requestValue` then back to a buffer with `valueBuffer`?

I'm trying to keep this backwards compatible with ldapjs and just be an updated drop in for anyone that might use ldapjs.

**Comment by dododedodonl** [comment]
I changed it so the original `ExtendedRequest.requestValue`'s behaviour is unchanged.

A new property, `requestValueBuffer` is used to read the string from the message. This is converted to a string, the same way `BerReader` does this, and stored in the original `requestValue`.

A new getter and setter `valueBuffer` is added, which returns the `requestValueBuffer` and sets the buffer to `value` which in turn sets it to `requestValue`.

In the constructor, `requestValueBuffer` is set to `requestValue` as a buffer, or and empty buffer if it is not defined.

File: lib/messages/ext_request.js [hunk 3]
```diff
@@ -49,22 +55,39 @@ Object.defineProperties(ExtendedRequest.prototype, {
...
+  valueBuffer: {
+    get: function getValueBuffer() {
+      return this.requestValueBuffer;
+    },
+    set: function setValueBuffer(val) {
+      if(!Buffer.isBuffer(val))
+        throw new TypeError('valueBuffer must be a buffer');
+
+      this.value = val;
+      this.requestValueBuffer = val;
```
Comment (UziTech) [inline lib/messages/ext_request.js]: `this.requestValueBuffer = val;` is handled with `this.value = val;`
```suggestion
```

File: test/messages/ext_request.test.js [hunk 4]
```diff
@@ -37,6 +61,7 @@ test('new with args', function (t) {
   t.ok(req);
   t.equal(req.requestName, '1.2.3.4');
   t.equal(req.requestValue, 'test');
+  t.ok(areBuffersEqual(req.requestValueBuffer, new Buffer('test', 'utf8')));
```
Comment (UziTech) [inline test/messages/ext_request.test.js]: We should probably be testing the public API too

```suggestion
  t.ok(areBuffersEqual(req.requestValueBuffer, new Buffer('test', 'utf8')));
  t.equal(req.value, 'test');
  t.ok(areBuffersEqual(req.valueBuffer, new Buffer('test', 'utf8')));
```

File: test/messages/ext_request.test.js [hunk 5]
```diff
@@ -51,6 +76,7 @@ test('parse', function (t) {
   t.ok(req._parse(new BerReader(ber.buffer)));
   t.equal(req.requestName, '1.2.3.4');
   t.equal(req.requestValue, 'test');
+  t.ok(areBuffersEqual(req.requestValueBuffer, new Buffer('test', 'utf8')));
```
Comment (UziTech) [inline test/messages/ext_request.test.js]: Same as above

```suggestion
  t.ok(areBuffersEqual(req.requestValueBuffer, new Buffer('test', 'utf8')));
  t.equal(req.value, 'test');
  t.ok(areBuffersEqual(req.valueBuffer, new Buffer('test', 'utf8')));
```

**Comment by dododedodonl** [comment]
Added some extra tests, also to test the setters

**Comment by UziTech** [comment]
🎉 Thanks for working through all the changes. I'll release a new version soon.

**Comment by dododedodonl** [comment]
> Thanks for working through all the changes. I'll release a new version soon.

Had to figure it out for my own setup, could as well share it with the rest of the world.

**Pull request merged**